core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -286,6 +286,44 @@ private[spark] class TaskSchedulerImpl(
     }
   }
 
+  /**
+   * SPARK-25250: Whenever any Task gets successfully completed, we simply mark the
+   * corresponding partition id as completed in all attempts for that particular stage and
+   * additionally, for a Result Stage, we also kill the remaining task attempts running on the
+   * same partition. As a result, we do not see any Killed tasks due to
+   * TaskCommitDenied Exceptions showing up in the UI. When this method is called from
+   * DAGScheduler.scala on a task completion event being fired, it is assumed that the new
+   * TaskSet has already been created and registered. However, a small possibility does exist
+   * that when this method gets called, possibly the new TaskSet might have not been added
```

Comment:
compared to #21131, I think the main point of this patch is to cover this corner case?

Reply:
Yes, indeed, this patch covers this corner case. But as @Ngone51 pointed out, there is an extremely rare chance that when this method gets called from DAGScheduler.scala, it could be possible that the new TaskSet might have not been added to taskSetsByStageIdAndAttempt. My assumption while writing this code was that there is always a small delay for the task completion event to propagate to the DAGScheduler, which gives sufficient time for the new TaskSet to be created gracefully. I have tested this patch like a million times and have not once encountered this case, but theoretically, there might be a small chance nevertheless. That is why I have added the comment, in case, somebody encounters this in the future.

Reply:
We really should try to be making this scenario impossible, not just less likely -- and in fact I think that *is* what this patch does.  The DAGScheduler does this by having everything that modifies shared state happen inside the event loop, which is single threaded.  The problem with the last fix was that it modified `TaskSchedulerImpl.taskSetsByStageIdAndAttempt`, but that does *not* have the same protections.  If we can ensure that the relevant data structures are only touched inside the event loop, we should be safe.  I don't want to just go on testing a million times, I've been burned by that before :P.  (Sorry I haven't had a chance to take a detailed walk through this patch yet though).

BTW I also meant earlier to change the *entire* PR description, not just the title, as well to that entire blurb I had.  That will become the commit msg, and its also helpful to any other reviewers.  I think that helps focus on what this is adding on top of the previous fix attempt, otherwise this sounds like its fixing the same thing:

SPARK-23433 tried to ensure that late task completions from a zombie taskset were properly updated in all tasksets for the stage (zombie or not). However, because it did this outside of the DAGScheduler event loop, the DAGScheduler could launch another taskset for the stage at the same time, before it had updated the set of tasks left to run. The new active taskset would never learn about the old completion from the zombie taskset. This could lead to multiple active tasksets (as described in SPARK-23433) or resultstage failure from repeated TaskCommitDenied exceptions for the task which the final stage attemps to re-run.

This change fixes it by moving that logic into the DAGScheduler event loop. As an optimization, it also kills tasks in all task sets, only if its a result stage (an extension of SPARK-25773).

Reply:
I think @squito has a good ponit here. Previously, I was thinking what if the active TaskSet has not been created when we marking completed partition for all TaskSets and does this fix still works ? Now, I realize that whether the active TaskSet has been created or not, it still works:

* created

obviously, fine.

* not created

then, when `DAGScheduler` calling `submitMissingTasks`, it will figure out which missing partitions to compute(including the partitions which were completed by tasks from previous stage attempt). So, the new created TaskSet also know about the completed partition. And these are all benefit from event loop, which perform as a single thread.

Reply:
Ahh ok, the event loop thread is taking care of all this. I see. Thank you. Have updated the PR description as well.